### PR TITLE
Handle chars literals in the parser for better syntax errors

### PR DESCRIPTION
**Pull request by whatyouhide**
This PR hopes to solve #4419.

Before this PR, we handled char literals (like `?a`) in the tokenizer, turning a literal like `?a` into the token `{:number, _, 97}` (thus indistinguishable from the literal `97` at the parsing stage). This led to error messages with the integer for the character instead of the character literal, e.g.:

```elixir
iex> :ok ?a
** (SyntaxError) iex:11: syntax error before: 97
```

With this PR, we now turn `?a` into the token `{:char, _, 97}` (which is the same token used by Erlang for Erlang char literals like `$a`); since it's the same token as in Erlang, the parser will now output the char literal as an Erlang char (`?a` would be printed as `$a`). We hijack the error message in `elixir_errors.erl` to end up with the correct message:

```elixir
iex> :ok ?a
** (SyntaxError) iex:11: syntax error before: ?a
```

@josevalim and I briefly discussed this on IRC and another possible solution came up: tokenize `?a` chars as `{'?', _, [$?, $a]}` tokens so that they'd be printed as `"?a"` in error messages and extracting the char integer in the parser would be a matter of getting the second element of the list. However, this doesn't play too nice with escaped chars (e.g., `?\s` would be printed as `"? "`) and also adds double quotes around the literal in parser error messages (which I personally don't like).

Let me know what you think and if this took the right direction after all :).

### CODE REVIEW

File: lib/elixir/test/erlang/tokenizer_test.erl [hunk 1]
```diff
@@ -147,12 +147,12 @@ space_test() ->
   [{op_identifier, {1,1,4}, foo}, {dual_op, {1,6,7}, '-'}, {number, {1,7,8}, 2}] = tokenize("foo  -2").
 
 chars_test() ->
-  [{number, {1,1,3}, 97}]      = tokenize("?a"),
-  [{number, {1,1,3}, 99}]      = tokenize("?c"),
-  [{number, {1,1,4}, 0}]       = tokenize("?\\0"),
-  [{number, {1,1,4}, 7}]       = tokenize("?\\a"),
-  [{number, {1,1,4}, 10}]      = tokenize("?\\n"),
-  [{number, {1,1,4}, 92}]      = tokenize("?\\\\").
+  [{char, {1,1,3}, 97}]      = tokenize("?a"),
```
Comment (lexmag) [inline lib/elixir/test/erlang/tokenizer_test.erl]: I'd removed this needless whitespace since this line will be included in blame anyway.
Same applies to lines after.

Comment (whatyouhide) [inline lib/elixir/test/erlang/tokenizer_test.erl]: I did that, then `push --force`'d over it and reinserted the whitespace because I expected you to say "this doesn't belong to this commit" :smile: Will do in a moment!

File: lib/elixir/src/elixir_tokenizer.erl [hunk 2]
```diff
@@ -175,9 +175,15 @@ tokenize([$~, S, H|T] = Original, Line, Column, Scope, Tokens) when ?is_sigil(H)
 
 % Char tokens
 
+% We tokenize char literals (?a) as {char, _, CharInt} instead of {number, _,
+% CharInt}. This is exactly what Erlang does with Erlang char literals
+% ($a). This means we'll have to adjust the error message for char literals in
+% the parser as by default {char, _, _} tokens are "hijacked" by Erlang and
```
Comment (lexmag) [inline lib/elixir/src/elixir_tokenizer.erl]: We're adjusting the error message in elixir_errors not parser. :bowtie:

File: lib/elixir/test/elixir/kernel/errors_test.exs [hunk 3]
```diff
@@ -830,6 +830,17 @@ defmodule Kernel.ErrorsTest do
...
   end
 
+  # As reported and discussed in
+  # https://github.com/elixir-lang/elixir/issues/4419.
+  test "characters literal are printed correctly in syntax errors" do
+    assert_compile_fail SyntaxError,
+      "nofile:1: syntax error before: ?a",
+      ':ok ?a'
+    assert_compile_fail SyntaxError,
+      "nofile:1: syntax error before: ?\\s",
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/errors_test.exs]: Let's also add a test for one unicode character with high value to ensure the "escaping" will also work as expected there and then we can :shipit:. :)

**Comment by whatyouhide** [comment]
@josevalim updated using `す` as the unicode character with a high value, let me know if it's ok.

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
